tests/api/products-attributes.php

```diff
@@ -0,0 +1,211 @@
+		wp_set_current_user( $this->user );
+		$request = new WP_REST_Request( 'GET', $this->endpoint . '/products' );
+		$request->set_param( 'attribute', 'pa_color' );
+		$request->set_param(
+			'attribute_term',
+			// Terms list needs to be a string.
+			$this->attr_term_ids['red'] . ',' . $this->attr_term_ids['yellow']
+		);
+
+		$response          = $this->server->dispatch( $request );
+		$response_products = $response->get_data();
+		$this->assertEquals( 200, $response->get_status() );
+		$this->assertEquals( 2, count( $response_products ) );
+	}
+
+	/**
+	 * Test getting products by multiple terms in one attribute, matching all.
+	 *
+	 * @since 1.2.0
+	 */
+	public function test_get_products_by_multiple_terms_all() {
+		wp_set_current_user( $this->user );
+		$request = new WP_REST_Request( 'GET', $this->endpoint . '/products' );
+		$request->set_param(
+			'attributes',
+			array(
+				'pa_color' => array(
+					$this->attr_term_ids['red'],
+					$this->attr_term_ids['blue'],
+				),
+			)
+		);
+		$request->set_param( 'attr_operator', 'AND' );
+
+		$response          = $this->server->dispatch( $request );
+		$response_products = $response->get_data();
+		$this->assertEquals( 200, $response->get_status() );
+		$this->assertEquals( 1, count( $response_products ) );
+	}
+
+	/**
+	 * Test getting products by multiple terms in multiple attributes, matching any.
+	 *
+	 * @since 1.2.0
+	 */
+	public function test_get_products_by_multiple_terms_multiple_attrs_any() {
+		wp_set_current_user( $this->user );
+		$request = new WP_REST_Request( 'GET', $this->endpoint . '/products' );
+		$request->set_param(
+			'attributes',
+			array(
+				'pa_color' => array( $this->attr_term_ids['red'] ),
+				'pa_size'  => array( $this->attr_term_ids['large'] ),
+			)
+		);
+		$request->set_param( 'attr_operator', 'IN' );
+
+		$response          = $this->server->dispatch( $request );
+		$response_products = $response->get_data();
+		$this->assertEquals( 200, $response->get_status() );
+		$this->assertEquals( 2, count( $response_products ) );
+	}
+
+	/**
+	 * Test getting products by multiple terms in multiple attributes, matching all.
+	 *
+	 * @since 1.2.0
+	 */
+	public function test_get_products_by_multiple_terms_multiple_attrs_all() {
+		wp_set_current_user( $this->user );
+		$request = new WP_REST_Request( 'GET', $this->endpoint . '/products' );
+		$request->set_param(
+			'attributes',
+			array(
+				'pa_color' => array( $this->attr_term_ids['blue'] ),
+				'pa_size'  => array( $this->attr_term_ids['medium'] ),
+			)
+		);
+		$request->set_param( 'attr_operator', 'AND' );
+
+		$response          = $this->server->dispatch( $request );
+		$response_products = $response->get_data();
+		$this->assertEquals( 200, $response->get_status() );
+		$this->assertEquals( 2, count( $response_products ) );
+	}
+
+	/**
+	 * Test getting products by attributes that don't exist.
+	 *
+	 * Note: This test is currently skipped because the API isn't registering the attribute
```

Comment:
Trying to follow along with this comment and what is failing in the code - so if i understand correctly it is `wc_get_attribute_taxonomy_names()` not returning the expected taxonomies when the API runs? Also should we just return an empty data set if an invalid attribute is sent?

In manual testing locally - the error message returned by the API is kind of confusing when sending an invalid attribute:

`"attributes": "pa_colorz is not a valid property of Object."`

Reply:
I'm not entirely sure what's going on, it seems to be a bug in the test environment– this works correctly when manually testing the API, but for some reason in the test, `wc_get_attribute_taxonomy_names` returns empty. Then the schema is not set up correctly, which originally prevented any valid attribute request. So I added [this workaround](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/7b571a412523b7ef7bbf7ef2e67eb1aa81ef3a44/includes/class-wgpb-products-controller.php#L341-L344), which removes the whitelist and lets in any attribute (even fake ones) when wc_get_attribute_taxonomy_names is empty.

Reply:
Yeah definitely confirmed it is working via manually sending requests. What do you think about the error message? Should it be more descriptive? "`attr` is not a valid product attribute" or something?

Reply:
We can't control that message, since [it's core validating against the schema](https://core.trac.wordpress.org/browser/tags/5.0/src/wp-includes/rest-api.php#L1144)

Reply:
TIL - thanks